File: rust/herb-analysis/src/ruby_locals_index/local.rs
use herb::location::Location;

#[derive(Debug, Clone, PartialEq)]
pub struct Local {
  pub name: String,
  pub declaration: Location,
  pub usages: Vec<Location>,
}

impl Local {
  pub fn new(name: impl Into<String>, declaration: Location, usages: Vec<Location>) -> Self {
    Self {
      name: name.into(),
      declaration,
      usages,
    }
  }

  pub fn locations(&self) -> Vec<Location> {
    std::iter::once(self.declaration).chain(self.usages.iter().copied()).collect()
  }
}

File: lib/herb/analysis/ruby_locals_index/local.rb
# frozen_string_literal: true

module Herb
  module Analysis
    class RubyLocalsIndex
      class Local
        attr_reader :name #: String
        attr_reader :declaration #: Herb::Location
        attr_reader :usages #: Array[Herb::Location]

        #: (String, Herb::Location, Array[Herb::Location]) -> void
        def initialize(name, declaration, usages)
          @name = name
          @declaration = declaration
          @usages = usages
        end

        #: () -> Array[Herb::Location]
        def locations
          [declaration, *usages]
        end
      end
    end
  end
end
